testcase/cross-region/tso.go

```diff
@@ -0,0 +1,250 @@
+package crossregion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
+	util2 "github.com/pingcap/tipocket/testcase/cross-region/pkg/util"
+)
+
+func (c *crossRegionClient) testTSO(ctx context.Context) error {
+	if err := c.requestTSOs(ctx); err != nil {
+		return err
+	}
+	log.Info("start to transfer Leader")
+	if err := c.transferLeader(); err != nil {
+		return err
+	}
+	log.Info("leader transfer committed")
+	if err := c.waitLeaderReady(); err != nil {
+		return err
+	}
+	log.Info("new leader ready")
+	for i := 1; i <= 3; i++ {
+		if err := c.transferPDAllocator(fmt.Sprintf("dc-%d", i)); err != nil {
+			return err
+		}
+	}
+	log.Info("new allocators ready")
+	// after transfer allocator and leader, we can tolerate tso failed once for each dc.	c.requestTSOAfterTransfer(ctx, "global")
+	c.requestTSOAfterTransfer(ctx, "dc-1")
+	c.requestTSOAfterTransfer(ctx, "dc-2")
+	c.requestTSOAfterTransfer(ctx, "dc-3")
+	return c.requestTSOs(ctx)
+}
+
+func (c *crossRegionClient) requestTSOs(ctx context.Context) error {
+	tsoCtx, tsoCancel := context.WithCancel(ctx)
+	defer tsoCancel()
+	tsoWg := &sync.WaitGroup{}
+	tsoWg.Add(4)
+	tsoErrCh := make(chan error, 4)
+	tsoCh := make(chan TSO, 4*c.TSORequests)
+	go c.requestTSO(tsoCtx, "global", tsoWg, tsoCh, tsoErrCh)
+	for i := 1; i <= 3; i++ {
+		go c.requestTSO(tsoCtx, fmt.Sprintf("dc-%v", i), tsoWg, tsoCh, tsoErrCh)
+	}
+	tsoWg.Wait()
+	if len(tsoErrCh) < 1 {
+		log.Info("requestTSOs success")
+		return nil
+	}
+	var tsoErrors []error
+	for len(tsoErrCh) > 0 {
+		tsoErr := <-tsoErrCh
+		tsoErrors = append(tsoErrors, tsoErr)
+	}
+	recordTSO := make(map[[2]int64]struct{}, 0)
```

Comment:
`recordTSO` only checks the uniqueness of TSO during this batch of requests. Considering the logical part has a big range to change, what about extending the check to a period of time such as 1 minute to cover more possibilities?

Reply:
updated.